src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -0,0 +1,315 @@
+                if (read != headerSize - INITIAL_HEADER_SIZE) {
+                    close();
+                    throw new RuntimeIOException(String.format("BCL %s has invalid header structure.", bclFile.getAbsoluteFile()));
+                }
+
+                byte bitsPerBasecall = headerBuffer.get();
+                byte bitsPerQualityScore = headerBuffer.get();
+
+                if (bitsPerBasecall != 2 && bitsPerBasecall != bitsPerQualityScore) {
+                    close();
+                    throw new RuntimeIOException("CBCL data not encoded in nibbles. (not currently supported)");
+                }
+
+                int numberOfBins = headerBuffer.getInt();
+
+                byte[] qualityBins = new byte[numberOfBins];
+                //each bin has a pair of 4 byte mappings
+                for (int j = 0; j < numberOfBins; j++) {
+                    int from = headerBuffer.getInt();
+                    int to = headerBuffer.getInt();
+                    qualityBins[j] = (byte) to;
+                }
+
+                int numTiles = headerBuffer.getInt();
+                TileData[] tileInfo = new TileData[numTiles];
+                for (int j = 0; j < numTiles; j++) {
+                    int tileNum = headerBuffer.getInt();
+                    int numClustersInTile = headerBuffer.getInt();
+                    int uncompressedBlockSize = headerBuffer.getInt();
+                    int compressedBlockSize = headerBuffer.getInt();
+                    tileInfo[j] = new TileData(tileNum, numClustersInTile, uncompressedBlockSize, compressedBlockSize);
+                }
+
+                boolean pfExcluded = headerBuffer.get() == 1;
+                cycleData[i] = new CycleData(version, headerSize, bitsPerBasecall, bitsPerQualityScore, numberOfBins, qualityBins, numTiles, tileInfo, pfExcluded);
+                this.streams[i] = stream;
+                this.streamFiles[i] = bclFile;
+                byteBuffer.clear();
+                headerBuffer.clear();
+            }
+
+
+        } catch (final IOException ioe) {
+            throw new RuntimeIOException(ioe);
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (queue == null || queue.size() == 0) {
+            advance();
+        }
+        return queue.size() != 0;
+    }
+
+    public BclData next() {
+        if (queue == null || queue.size() == 0) {
+            advance();
+        }
+
+        final BclData data = queue.get(0);
+        queue.remove(0);
+        return data;
+    }
+
+    @Override
+    public void close() {
+        for (final InputStream stream : this.streams) {
+            CloserUtil.close(stream);
+        }
+    }
+
+    private void advance() {
+        int totalCycleCount = 0;
+        BclData data = new BclData(outputLengths);
+
+        for (int read = 0; read < outputLengths.length; read++) {
+            for (int cycle = 0; cycle < outputLengths[read]; ++cycle) {
```

Comment:
post increment in one line and pre in the next.  Potentially confusing.  Maybe be consistent and use a <= in one of these for loops.  Or else comment?